### DISCUSSION TITLE
How to logout Hydra and Kratos at the same time

### DISCUSSION THREAD

**Discussion by EiffelFly** [ory/hydra#2852] [Q&A] [answered]
Hi, Ory community.

First of all, thanks your hard work for providing this open source solution. Our project really benefit a lot from it due to we need to issue lots of identities to users.

Recently, we are approaching logout flow, and I feel confused toward how to logout Hydra and Kratos at the same time.

The flow we are using is OAuth with github as OIDC provider. Once user click logout button, we would like to log out all the session related to not just Hydra but also Kratos. As far as I know the flow will be:

1. Redirect browser to /logout, because we don't have logout_challenge at the point, we move on to next part.
2. Redirect browser to the hydra `/oauth2/sessions/logout` endpoint with idToken if SSR.
3. If the session is valid, this endpoint will redirect browser to the /logout page which set by hydra configuration with logout_challenge query parameter.
4. Once we have logout_challenge we can use `acceptLogoutRequest` sdk then redirect browser to the /oauth2/sessions/logout?logout_verifier=?

Now we had successfully wiped out oauth2_authentication_session cookie, here comes my question: How do we wipe out Kratos session from this point.

Thanks again!

**Comment by vinckr** [comment, marked as answer]
Not knowing much about your use case, 
is revoking the option with the [Self-Service Logout for API Clients](https://www.ory.sh/kratos/docs/next/self-service/flows/user-logout/#self-service-logout-for-api-clients) an option?

**Reply by EiffelFly** [reply]
Thanks for the quick response!!!

I have found a solution, but i would like to know whether this pattern is good.

We set Hydra env config URLS_POST_LOGOUT_REDIRECT=http://127.0.0.1:3000/logout?logout_kratos=true

1. Redirect browser to /logout, because we don't have logout_challenge at the point, we move on to next part.
2. Redirect browser to the hydra /oauth2/sessions/logout endpoint with idToken if SSR.
3. If the session is valid, this endpoint will redirect browser to the /logout page which set by hydra configuration with logout_challenge query parameter.
4. Once we have logout_challenge we can use acceptLogoutRequest sdk then redirect browser to the /oauth2/sessions/logout?logout_verifier=?
5. After redirect to step 4 url, Hydra will logout cookie for us and redirect to the URLS_POST_LOGOUT_REDIRECT url, we have a parameter which indicate logout page to execute the logout flow @vinckr just mentioned

Now we can logout both Kratos and Hydra all within same process(in user point of view)

Or we can actually pass `<post_logout_redirect_uris>` when we access `/oauth2/sessions/logout` endpoint?

**Reply by EiffelFly** [reply]
In order to make <post_logout_redirect_uris> accept /logout?logout_kratos=true, we have to whitelist it in client config, but when I try to do this the error show no matches found: http://127.0.0.1:3000/logout?logout_kratos=true/

```
docker-compose exec hydra \
    hydra clients create \
    --endpoint http://127.0.0.1:4445 \
    --id test \
    --secret secret \
    --name Instill CLI test \
    --client-uri http://127.0.0.1:3000 \
    --policy-uri http://127.0.0.1:3000/policy \
    --grant-types authorization_code,refresh_token \
    --response-types code,id_token \
    --scope openid,offline,email \
    --audience https://api.test.com \
    --callbacks http://127.0.0.1:3000/callback \
    --post-logout-callbacks http://127.0.0.1:3000/logout?logout_kratos=true
```

**Reply by atreya2011** [reply]
@EiffelFly You have proposed a very interesting solution! I have yet to resolve this issue in my Kratos + Hydra integration. I will try this out and let you know how it goes (or maybe another solution if I bump into the same problem as you did).

**Reply by vinckr** [reply]
@EiffelFly 
The Ory Kratos and Ory Hydra integration just landed with Hydra 2.0, 
see the [release notes](https://github.com/ory/hydra/releases/tag/v2.0.0) for more details!

**Reply by EiffelFly** [reply]
@vinckr Awesome!! Really looks forward to implementing these new features. Great work!

**Reply by NhutTran98** [reply]
Hi @EiffelFly, I have tried your solution but I found that endpoint `http://127.0.0.1:3000/logout?logout_kratos=true` can be attacked by csrf. @vinckr I also try API client but I don't find a way to make it work with web browser. I think we need a way for admin to revoke user session.

**Reply by vinckr** [reply]
I recommend to use [Ory Kratos](https://github.com/ory/kratos/) for that, it has full user management APIs and also integrates with Hydra now, see [above](https://github.com/ory/hydra/discussions/2852#discussioncomment-4019642).

**Reply by jtc3161** [reply]
@vinckr Looking at the source code, kratos answers the login_challenge, but not the logout_challenge. Is this a missing feature or is there a reason behind this? I imagine not everyone would want to be logged out of kratos, specifically third part clients, but it does seem to me that the included kratos+hydra integration does not resolve this discussion at the moment.

**Reply by vinckr** [reply]
Hm AFAICT there are multiple session layers with OAUth2:

- Ory Identities/Kratos keeps track of the user within your application.
- Ory OAuth2 & OpenID Connect maintains a session cookie. When a user signs in, the cookie will be set for that user. This allows the next OAuth2 request to complete without requesting the user to sign in again.

So when the user logs out, you usually only want to log out of the Ory Identities session, but the OAuth2 session should still be there (to complete the next log in without going through the consent stage again). 
Might be that I got something wrong 🤔 . Feel free to open a dedicated discussion, if this makes it not clearer.

**Reply by sidharthramesh** [reply]
@vinckr take a scenario where the user wants to log out of Kratos, and then log back in with another identity.
The OAuth2 flow errors out stating that the subject_id of the previously set cookie does not match the current one initiating the flow.

Even if the identity stays the same, I don't think the user would expect an OAuth2 flow to progress without a login page since they have "signed out".


I'm using Kratos + Hydra together. I am facing issues when launching OAuth2 requests when the logged-in user changes in Kratos, but not in Hydra. Given that there has been a lot of development around Hydra - Kratos integration with the release of Kratos 1.0, what is the best way to simultaneously reset the Hydra login session when the Kratos user logs out?

**Reply by atreya2011** [reply]
@sidharthramesh 
Maybe this will be helpful. During the logout flow, I revoke the hydra logout session, consent session and along with the logout session in kratos. 
https://github.com/atreya2011/go-kratos-test/blob/062f7ab9323b1c4ad559f4101a20cdda9cf5f2cd/main.go#L188

@vinckr 
I finally found the time to update my example implementation to the latest version of Kratos and Hydra :)

**Reply by vinckr** [reply]
>take a scenario where the user wants to log out of Kratos, and then log back in with another identity.

I agree, if that is part of the use case then you need to take it into account. 
Maybe evaluate if you even need OAuth2 for your use case, then this would not be an issue.
Read more about this here: https://www.ory.sh/oauth2-openid-connect-do-you-need-use-cases-examples/ and especially the part about sessions.

**Reply by sidharthramesh** [reply]
Hey @atreya2011 and @vinckr, Thank you so much for your input.
